### ISSUE TITLE
MNIST Benchmark vs Pytorch

### ISSUE THREAD

**Issue by pedronahum** [pedronahum/TaylorTorch#6]
Metric | Python | Swift | Ratio
-- | -- | -- | --
Avg Forward | 29.0ms | 16.2ms | Swift 1.8x faster ✅
Avg Backward | 36.5ms | 213.1ms | Swift 5.8x slower ❌
Avg Optimizer | 0.3ms | 1.6ms | Swift 5x slower

<br class="Apple-interchange-newline">

The problem is in [Sequential.swift:188] in the Chain._vjpCallAsFunction method:

```swift
@derivative(of: callAsFunction, wrt: (self, x))
public func _vjpCallAsFunction(_ x: Input)
  -> (value: Output, pullback: (Output.TangentVector) -> (TangentVector, Input.TangentVector))
{
  // Forward through first, then second.
  let y1 = first(x)                                        // Forward pass 1
  let (y, pbSecond) = second.appliedForBackpropagation(to: y1)  // Forward pass 2
  return (
    y,
    { v in
      let (dSecond, dY1) = pbSecond(v)
      let (_, pbFirst) = first.appliedForBackpropagation(to: x)  // ❌ REDUNDANT FORWARD PASS!
      let (dFirst, dX) = pbFirst(dY1)
      return (TangentVector(first: dFirst, second: dSecond), dX)
    }
  )
}
```

The bug: Inside the pullback closure, first.appliedForBackpropagation(to: x) is called again, which performs another forward pass just to get the pullback function. This forward pass was already done at line 181 (let y1 = first(x)). Impact: For a Sequential with N layers (nested as Chain<Chain<Chain<...>>>), this creates: 1) During forward: N forward passes (correct) and 2)During backward: N + (N-1) + (N-2) + ... + 1 = O(N²) forward passes (not good!)
 
The MNIST model has ~11 layers, so backward pass does ~66 extra forward passes instead of 0.

**Comment by pedronahum** [comment]
A way to address this issue:

The pullback for first should be captured during the forward pass, not inside the pullback closure:

```swift
@derivative(of: callAsFunction, wrt: (self, x))
public func _vjpCallAsFunction(_ x: Input)
  -> (value: Output, pullback: (Output.TangentVector) -> (TangentVector, Input.TangentVector))
{
  // Capture both forward outputs AND pullbacks during forward pass
  let (y1, pbFirst) = first.appliedForBackpropagation(to: x)    // ✅ Get pullback here
  let (y, pbSecond) = second.appliedForBackpropagation(to: y1)
  return (
    y,
    { v in
      let (dSecond, dY1) = pbSecond(v)
      let (dFirst, dX) = pbFirst(dY1)  // ✅ Use captured pullback
      return (TangentVector(first: dFirst, second: dSecond), dX)
    }
  )
}
```


Metric | Swift (Before) | Swift (After) | Python | Improvement
-- | -- | -- | -- | --
Avg Backward | 213.1ms | 62.7ms | 33.0ms | 3.4x faster
Avg Forward | 13.6ms | 13.4ms | 26.4ms | Swift is 2x faster
Avg Optimizer | 1.3ms | 1.3ms | 0.3ms | -

<p style="display: inline; color: rgb(204, 204, 204); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

The backward pass went from<span> </span><strong>5.8x slower</strong><span> </span>than Python to only<span> </span><strong>1.9x slower</strong>.</p>

**Comment by pedronahum** [comment]
## Native BatchNorm Implementation for Improved Backward Pass Performance

### Summary
- Integrate native ATen `batch_norm` and `batch_norm_backward` kernels into the BatchNorm layer
- Achieved **30% improvement** in backward pass performance (68ms → 47ms)
- Total training compute time improved by **25%** (8.35s → 6.23s for 100 batches)

### Benchmark Results (100 batches, batch_size=128)

| Metric | Before | After | Python | 
|--------|--------|-------|--------|
| Forward | 13.9ms | 13.4ms | 27.4ms |
| Backward | 68.0ms | **47.4ms** | 33.1ms |
| Optimizer | 1.65ms | 1.41ms | 0.26ms |
| Total Compute | 8.35s | **6.23s** | 6.07s |

### Files Changed

**Sources/ATenCXX/include/tensor_shim.hpp**
- Added C++ bindings for `_native_batch_norm` and `_native_batch_norm_backward`
- Added tuple getter functions for extracting results

**Sources/Torch/ATen/Core/Tensor/Tensor+NN.swift**
- Added `nativeBatchNorm()` Swift wrapper for forward pass
- Added `_nativeBatchNormBackward()` Swift wrapper for backward pass

**Sources/Torch/Modules/Layers/BatchNorm.swift**
- Modified `callAsFunction` to use native batch norm in training mode
- Added `_nativeBatchNormImpl` with custom VJP using native backward kernel

### Technical Notes
- Swift AD requires `valueWithPullback` pattern in VJPs - direct pullback returns are ignored
- Control flow (switch/if) inside VJP primal functions can prevent custom derivatives from being selected
- Native ATen kernels avoid tensor allocation overhead and use optimized BLAS routines

### Test Plan
- [x] MNIST training runs correctly with expected accuracy improvement
- [x] Release build benchmarks show consistent ~47ms backward pass
- [x] No regressions in forward pass performance

**Comment by JaapWijnen** [comment]
Is there an easy way to run this comparison and compare the code? I'd love to see

**Comment by pedronahum** [comment]
> Is there an easy way to run this comparison and compare the code? I'd love to see

please have a look at: https://github.com/pedronahum/TaylorTorch/tree/performance_improvements; there I have been bechmarking the code on my mac

**Comment by JaapWijnen** [comment]
How do I run those benchmarks? (I'm not sure where to look for them on that branch)

**Comment by pedronahum** [comment]
> How do I run those benchmarks? (I'm not sure where to look for them on that branch)

Right now it is a little bit painful, as we need to compile pytorch from source. Example of how I do it in my mac below. Code is in the Examples folder. 

# Swift benchmark
cd /Users/pedro/programming/swift/TaylorTorch
PYTORCH_INSTALL_DIR="/Users/pedro/programming/pytorch/install" \
CXXFLAGS="-I/Users/pedro/programming/pytorch/install/include -I/Users/pedro/programming/pytorch/install/include/torch/csrc/api/include" \
swift build -c release
.build/release/MNISTExample --epochs=3

# Python benchmark
cd Examples/MNIST/python
pip install -r requirements.txt
python main.py --epochs=3 --device=cpu

**Comment by thinkpractice** [comment]
@JaapWijnen @pedronahum if you don't want to build `pytorch` from scratch, you can use a pre-built version of `libtorch`. I've added two scripts for that [here](https://github.com/thinkpractice/TaylorTorch/tree/linux_fixes).

- [get-pre-build-torch.sh](https://github.com/thinkpractice/TaylorTorch/blob/linux_fixes/get-pre-build-torch.sh) downloads the correct libtorch for Linux, you may need to adjust this if your building on mac
- [setenv.sh](https://github.com/thinkpractice/TaylorTorch/blob/linux_fixes/setenv.sh) sets the environment for the build. It configures it in such a way that the pre-build `libtorch` is used.

I have tested this on Linux and it was working. We don't know however if a `libtorch` built with `gcc` would cause some incompatibilities or other issues. That's why we opted to build `libtorch` from scratch with `clang`.

**Comment by thinkpractice** [comment]
Hi @pedronahum,

I also ran your benchmark on my PC with a `Intel® Core™ i7-8700K × 12` with 32GB ram running `Ubuntu 25.10`

| Metric | Python  | Swift | Swift Docker |
| --------   | ---------    | ------- | -----------------  |
Version  | 3.11.14 | main-snapshot-2025-11-03 | main-snapshot-2025-11-03 |
LibTorch version | 2.9.1+cu128 | 2.8.0 (pre-built) | 2.8.0 |
Avg forward pass time:    |  41.4574ms  |  17.8221ms |  32.5629ms |
Avg backward pass time: |   23.1108ms | 248.2159ms | 630.7500ms |
Avg optimizer time:           |  0.5258ms   | 1.8207ms | 10.7680ms |
Total compute time:         |  91.587s  | 376.877s |   948.432s |



Note the backward pass is really slow, but comparable to your first benchmark. I guess the changes you've made are not on the main branch yet? This is with the latest main branch merged. I also had to solve a crash with `Data(contentsOf:)` not working on Linux with Swift 6.3.1. I think this is a bug in the new `swift-corelibs-foundation`.

We need to be careful that we're not comparing apples and oranges here as the Python uses the latest libtorch 2.9.1 and Swift uses 2.8.0. Do you think we can update the swift libtorch to 2.9.1 as well? I will also try running everything in the docker. Let's see how this goes.

kind regards,

Tim

**Comment by pedronahum** [comment]
> We need to be careful that we're not comparing apples and oranges here as the Python uses the latest libtorch 2.9.1 and Swift uses 2.8.0. Do you think we can update the swift libtorch to 2.9.1 as well? I will also try running everything in the docker. Let's see how this goes.

Hi @thinkpractice - sorry if I was not clear but I used the exact same library compiled from source (2.8) to avoid noise. Swift and Python are being benchmark above with the same optimization flags and compiler. Need to update all the install scripts to make it easy to follow / replicate.

**Comment by thinkpractice** [comment]
Hi @pedronahum, ah good! I was thinking it might be comparing different versions. Running TaylorTorch benchmark with libtorch 2.9.1 now. Let's see :)

**Comment by pedronahum** [comment]
My goal is to push tomorrow all the changes from tagged version to the main repo. Some changes are required as I was only testing on my mac. But I am on it.
